go-controller/pkg/ovn/controller/egress_services/egress_services_controller.go

```diff
@@ -0,0 +1,365 @@
+// In addition, it removes the egress service labels of deleted services from the nodes.
+func (c *Controller) repair() error {
+	c.Lock()
+	defer c.Unlock()
+
+	svcKeyToAllV4Endpoints := map[string]sets.String{}
+	svcKeyToAllV6Endpoints := map[string]sets.String{}
+	svcKeyToConfiguredV4Endpoints := map[string][]string{}
+	svcKeyToConfiguredV6Endpoints := map[string][]string{}
+	services, _ := c.serviceLister.List(labels.Everything())
+
+	for _, svc := range services {
+		if util.HasEgressSVCAnnotation(svc) && util.HasEgressSVCHostAnnotation(svc) &&
+			util.ServiceTypeHasLoadBalancer(svc) && len(svc.Status.LoadBalancer.Ingress) > 0 {
+			var err error
+			key, _ := cache.MetaNamespaceKeyFunc(svc)
+			conf, err := util.ParseEgressSVCAnnotation(svc)
+			if err != nil && !util.IsAnnotationNotSetError(err) {
+				klog.Errorf("can't parse %s egress service configuration, err: %v", key, err)
+				continue
+			}
+			svcHost, _ := util.GetEgressSVCHost(svc)
+			nodeState, ok := c.nodes[svcHost]
+			if !ok {
+				nodeState, err = c.nodeStateFor(svcHost)
+				if err != nil {
+					klog.Errorf("can't fetch egress service %s node %s state, err: %v", key, svcHost, err)
+					continue
+				}
+			}
+
+			v4, v6, epNodes, err := c.allEndpointsFor(svc)
+			if err != nil {
+				klog.Errorf("can't fetch all endpoints for egress service %s, err: %v", key, err)
+				continue
+			}
+
+			totalEps := len(v4) + len(v6)
+			if totalEps == 0 {
+				klog.Errorf("Egress service %s has no endpoints", key)
+				continue
+			}
+
+			hostValid := true
+			if svc.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeLocal {
+				hostValid = false
+				for _, node := range epNodes {
+					if node == svcHost {
+						hostValid = true
+						break
+					}
+				}
+			}
+
+			if !hostValid {
+				klog.Errorf("Node %s no longer valid for egress service %s", svcHost, key)
+				continue
+			}
+
+			svcKeyToAllV4Endpoints[key] = v4
+			svcKeyToAllV6Endpoints[key] = v6
+			svcKeyToConfiguredV4Endpoints[key] = []string{}
+			svcKeyToConfiguredV6Endpoints[key] = []string{}
+
+			selector, _ := metav1.LabelSelectorAsSelector(&conf.NodeSelector)
+			svcState := &svcState{node: svcHost, selector: selector, v4Endpoints: sets.NewString(), v6Endpoints: sets.NewString(), stale: false}
+			nodeState.allocations[key] = svcState
+			c.nodes[svcHost] = nodeState
+			c.services[key] = svcState
+		}
+	}
+
+	p := func(item *nbdb.LogicalRouterPolicy) bool {
+		if item.Priority != ovntypes.EgressSVCReroutePriority {
+			return false
+		}
+
+		svcKey, found := item.ExternalIDs["EgressSVC"]
+		if !found {
+			klog.Infof("Egress service repair will delete lrp because it uses the egress service priority but does not belong to one: %v", item)
+			return true
+		}
+
+		svc, found := c.services[svcKey]
+		if !found {
+			klog.Infof("Egress service repair will delete %s because it is no longer a valid egress service: %v", svcKey, item)
+			return true
+		}
```

Comment:
BTW I am not sure if we should be bothering with this but this is the perfect use case for adding a secondary index as client index so that looking up policies with `ExternalIDs["EgressSVC"]==svcKey` is fast.

https://github.com/ovn-org/libovsdb#client-indexes